review-agent/repowatch/controllers/repowatch_controller.go

```diff
@@ -0,0 +1,322 @@
+		}
+	}
+
+	// Create new sandboxes
+	for _, pr := range prs {
+		sandboxName := fmt.Sprintf("%s-pr-%d", strings.Split(repoWatch.Spec.RepoURL, "/")[len(strings.Split(repoWatch.Spec.RepoURL, "/"))-1], *pr.Number)
+		sandboxExists := false
+		for _, sandbox := range sandboxes.Items {
+			if sandbox.GetName() == sandboxName {
+				sandboxExists = true
+				// Check if replica count > 0
+				replicas, found, err := unstructured.NestedInt64(sandbox.Object, "spec", "replicas")
+				if err != nil || !found {
+					log.Error(err, "unable to get replicas for sandbox", "sandbox", sandbox.GetName())
+					break
+				}
+				if replicas > 0 {
+					activeSandboxes++
+				}
+				watchedPRs = append(watchedPRs, reviewv1alpha1.WatchedPR{
+					Number:      *pr.Number,
+					SandboxName: sandboxName,
+					Status:      "Active",
+				})
+				break
+			}
+		}
+
+		if !sandboxExists {
+			if activeSandboxes < repoWatch.Spec.MaxActiveSandboxes {
+				log.Info("creating sandbox for pr", "pr", *pr.Number)
+				if err := r.createSandboxForPR(ctx, repoWatch, pr); err != nil {
+					log.Error(err, "unable to create sandbox for pr", "pr", *pr.Number)
+				} else {
+					activeSandboxes++
+					watchedPRs = append(watchedPRs, reviewv1alpha1.WatchedPR{
+						Number:      *pr.Number,
+						SandboxName: sandboxName,
+						Status:      "Creating",
+					})
+				}
+			} else {
+				pendingPRs = append(pendingPRs, reviewv1alpha1.PendingPR{
+					Number: *pr.Number,
+					Status: "Pending",
+				})
+			}
+		}
+	}
+
+	repoWatch.Status.ActiveSandboxCount = activeSandboxes
+	repoWatch.Status.WatchedPRs = watchedPRs
+	repoWatch.Status.PendingPRs = pendingPRs
+
+	return r.Status().Update(ctx, repoWatch)
+}
+
+func (r *RepoWatchReconciler) generatePullRequestPrompt(repoWatch *reviewv1alpha1.RepoWatch, pr *github.PullRequest) (string, error) {
+	promptTmpl := "You are an expert kubernetes developer who is helping with code reviews. Please look at the PR {{.Number}} linked at {{.HTMLURL}} provide a review feedback."
+	if repoWatch.Spec.Prompt != "" {
+		promptTmpl = repoWatch.Spec.Prompt
+	}
+	tmpl, err := template.New("myTemplate").Parse(promptTmpl)
+	if err != nil {
+		return "", err
+	}
+
+	// Create a bytes.Buffer to capture the template output
+	var buf bytes.Buffer
+	// Execute the template, writing the output to the buffer
+	err = tmpl.Execute(&buf, pr)
+	if err != nil {
+		return "", err
+	}
+
+	// Get the rendered string from the buffer
+	prompt := buf.String()
+
+	return prompt, nil
+}
+
+func (r *RepoWatchReconciler) createSandboxForPR(ctx context.Context, repoWatch *reviewv1alpha1.RepoWatch, pr *github.PullRequest) error {
+	log := log.FromContext(ctx)
+	repoName := strings.Split(repoWatch.Spec.RepoURL, "/")[len(strings.Split(repoWatch.Spec.RepoURL, "/"))-1]
+	sandboxName := fmt.Sprintf("%s-pr-%d", repoName, *pr.Number)
+
+	prompt, err := r.generatePullRequestPrompt(repoWatch, pr)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Generated sandbox for PR", "pr", *pr)
```

Comment:
I do wonder whether we would better off having one controller that creates a CRD for each open github PR, and then we could just use KRO to do this ... but let's see!